Exclude If Present

shivansh02 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

### Description
<!--- Describe your changes in detail -->
Exclude directories where specified files are present.

### Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
#1863 

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Tested manually by running backups. Selected filenames get appended to borg create command as "--exclude-if-present filename". Modified `test_exclusion_preview_populated` to test preview tab.
### Screenshots (if appropriate):
![Screenshot from 2024-06-12 18-47-02](https://github.com/borgbase/vorta/assets/89853707/e594502e-ca3d-4087-a63c-a4d7c80b6762)
![Screenshot from 2024-06-12 18-47-08](https://github.com/borgbase/vorta/assets/89853707/2306f311-81a8-4a49-afad-f24b9bde48be)

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [CONTRIBUTING](https://vorta.borgbase.com/contributing/) guide.
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.


*I provide my contribution under the terms of the [license](./../LICENSE.txt) of this repository and I affirm the [Developer Certificate of Origin][dco].*

[dco]: https://developercertificate.org/

<!--
This template is sourced from the awesome https://github.com/TalAter/open-source-templates
-->

m3nu commented:
Testing this locally now. In the future, you can also rebase a new PR if there were commits in the meantime.

m3nu commented:
Works as expected locally and passes the correct args to Borg.

I would still suggest to add some common example files in the dialog, but keep them unselected. Like the ones shown in Borg's docs.

shivansh02 commented:
> I would still suggest to add some common example files in the dialog, but keep them unselected. Like the ones shown in Borg's docs.

I added two files (.nobackup and .vortaignore). Did these not show up for you?
I added the files as default value for exclude_if_present field in models.py and wrote a migration for the same. Did I miss something?

m3nu commented on src/vorta/store/models.py:
```diff
@@ -79,7 +79,7 @@ class BackupProfileModel(BaseModel):
     compression = pw.CharField(default='lz4')
     exclude_patterns = pw.TextField(null=True)
-    exclude_if_present = pw.TextField(null=True)
+    exclude_if_present = pw.TextField(null=True, default="[] .nobackup\n[] .vortaignore")
```
`.nobackup` is fine, but I wouldn't introduce a new standard, like `.vortaignore`. Why not add `CACHEDIR.TAG`, which is an existing [standard](https://bford.info/cachedir/) to support the related Borg flag.

shivansh02 replied:
Wasn't aware about CACHEDIR.TAG, added it now.

m3nu commented on src/vorta/store/migrations.py:
```diff
@@ -250,6 +250,17 @@ def run_migrations(current_schema, db_connection):
+                migrator.add_column_default(
+                    BackupProfileModel._meta.table_name,
+                    'exclude_if_present',
+                    pw.TextField(default="[] .nobackup\n[] .vortaignore"),
```
See related comment for model

m3nu commented on src/vorta/assets/UI/excludedialog.ui:
```diff
@@ -137,6 +177,7 @@
        </item>
       </layout>
      </widget>
+
```
Why are we adding a new line here?

m3nu commented on src/vorta/views/exclude_dialog.py:
```diff
@@ -310,6 +352,73 @@ def preset_item_changed(self, item):
+            else:
+                item.setText(pattern)
+                item.setCheckable(True)
+                item.setCheckState(Qt.CheckState.Unchecked)
```
Doesn't this mean that all existing exclude settings the user had are now disabled?

shivansh02 replied:
@m3nu That's true, I didn't take the existing ones into account. Should I try to make them checked by default?

m3nu replied:
Yes, I would check them by default.

m3nu commented:
I think we're good now. Let me run it again locally tonight.
